### PR TITLE
Add 4.0 and fix release notes for Head

### PR DESCRIPTION
**Pull request by juliogonzalez**
- Add 4.0 everywhere for nightlies and released, for server, proxy, and client tools, and retail product and module for the future.
- Add the ToSLE Head subproject, wich contains now the release notes. We have then on a separate project to make the submissions easier, as release-notes are to be submitted directly to SLE.

Verify:
- [x] 4.0-nightly, without mirror
- [x] 4.0-released, without mirror
- [x] 4.0-nightly, with mirror
- [x] 4.0-released, with mirror

### CODE REVIEW

**Comment by juliogonzalez** [comment]
@moio I have no idea of what the failure means, can you provide any hint?

File: salt/repos/suse_manager_server.sls [hunk 1]
```diff
@@ -130,6 +163,19 @@ suse_manager_devel_repo:
...
 
+{% if '4.0-nightly' in grains['product_version'] %}
+suse_manager_devel_repo:
+  file.managed:
+    - name: /etc/zypp/repos.d/Devel_Galaxy_Manager_4.0.repo
+    - source: salt://repos/repos.d/Devel_Galaxy_Manager_4.0.repo
+    - template: jinja
+
+suse_manager_devel_releasenotes_repo:
+    - name: /etc/zypp/repos.d/Devel_Galaxy_Manager_Head_ToSLE.repo
```
Comment (mcalmer) [inline salt/repos/suse_manager_server.sls]: I think here we miss a `file.managed` .

File: salt/repos/suse_manager_proxy.sls [hunk 2]
```diff
@@ -113,6 +146,19 @@ suse_manager_devel_repo:
...
 
+{% if '4.0-nightly' in grains['product_version'] %}
+suse_manager_devel_repo:
+  file.managed:
+    - name: /etc/zypp/repos.d/Devel_Galaxy_Manager_4.0.repo
+    - source: salt://repos/repos.d/Devel_Galaxy_Manager_4.0.repo
+    - template: jinja
+
+suse_manager_devel_releasenotes_repo:
+    - name: /etc/zypp/repos.d/Devel_Galaxy_Manager_4.0_ToSLE.repo
```
Comment (mcalmer) [inline salt/repos/suse_manager_proxy.sls]: Here we miss a `file.managed`.

File: salt/repos/suse_manager_server.sls [hunk 3]
```diff
@@ -62,6 +88,13 @@ suse_manager_pool_repo:
     {% endif %}
     - template: jinja
 
+{% if grains['osfullname'] != 'Leap' %}
+suse_manager_devel_releasenotes_repo:
+    - name: /etc/zypp/repos.d/Devel_Galaxy_Manager_Head_ToSLE.repo
```
Comment (mcalmer) [inline salt/repos/suse_manager_server.sls]: I think here we miss a `file.managed` .

File: salt/repos/suse_manager_proxy.sls [hunk 4]
```diff
@@ -62,6 +88,13 @@ suse_manager_proxy_pool_repo:
     {% endif %}
     - template: jinja
 
+{% if grains['osfullname'] != 'Leap' %}
+suse_manager_devel_releasenotes_repo:
+    - name: /etc/zypp/repos.d/Devel_Galaxy_Manager_Head_ToSLE.repo
```
Comment (juliogonzalez) [inline salt/repos/suse_manager_proxy.sls]: And here `file.managed` is missing as well.

File: salt/repos/suse_manager_server.sls (hunk 1 above)
Comment (juliogonzalez) [reply salt/repos/suse_manager_server.sls]: Fixed now.

File: salt/mirror/minima.yaml [hunk 5]
```diff
@@ -56,13 +56,26 @@ scc:
     - SUSE-Manager-Server-3.1-Updates
     - SUSE-Manager-Server-3.2-Pool
     - SUSE-Manager-Server-3.2-Updates
+    - SLE-Product-SUSE-Manager-Server-Pool
```
Comment (juliogonzalez) [inline salt/mirror/minima.yaml]: To be honest, I am not sure about the repository names for the new products and modules so... how can I check it?

Comment (moio) [reply salt/mirror/minima.yaml]: Run minima manually (with a mirror host or just by downloading the binary and run it on your machine). It will output a list of available repos before starting syncing anything, and you can copypaste from there.

File: README_ADVANCED.md [hunk 6]
```diff
@@ -9,13 +9,15 @@ Some modules have a `product_version` variable that determines the software prod
 Legal values for released software are:
  * `3.0-released`   (latest released Maintenance Update for SUSE Manager 3.0 and Tools)
  * `3.1-released`   (latest released Maintenance Update for SUSE Manager 3.1 and Tools)
- * `3.2-released`   (latest released alpha/beta/gold master candidate for SUSE Manager 3.2 and Tools)
+ * `3.2-released`   (latest released Maintenance Update for SUSE Manager 3.2 and Tools)
+ * `4.0-released`   (latest released Maintenance Update for SUSE Manager 3.2 and Tools)
```
Comment (moio) [inline README_ADVANCED.md]:
```suggestion
 * `4.0-released`   (latest released Maintenance Update for SUSE Manager 4.0 and Tools)
```

File: modules/libvirt/pts/main.tf [hunk 7]
```diff
@@ -2,17 +2,17 @@ module "server" {
   source = "../suse_manager"
   name = "${var.suse_manager_name}"
   base_configuration = "${var.base_configuration}"
-  product_version = "3.2-nightly"
-  image = "sles12sp3"
+  product_version = "4.0-nightly"
```
Comment (moio) [inline modules/libvirt/pts/main.tf]: Please separate out the PTS changes to a separate PR, we have to review and commit them on their own as it will impact our running testsuite.

File: salt/mirror/minima.yaml [hunk 8]
```diff
@@ -135,6 +148,10 @@ http:
   - url: http://download.suse.de/ibs/Devel:/Galaxy:/Manager:/3.2/SLE_12_SP3
     archs: [x86_64]
 
+  # SUSE Manager 4.0 devel
+  - url: http://download.suse.de/ibs/Devel:/Galaxy:/Manager:/4.0/SLE_12_SP3
```
Comment (moio) [inline salt/mirror/minima.yaml]: Is the last part of the URL (`SLE_12_SP3`) really correct?

File: salt/repos/repos.d/Devel_Galaxy_Manager_4.0_ToSLE.repo [hunk 9]
```diff
@@ -0,0 +1,6 @@
+[Devel_Galaxy_Manager_4.0_ToSLE]
+name=Devel Project for SUSE Manager Head (packages submitted directly to SLE)
+type=rpm-md
+enabled=1
+baseurl=http://{{ grains.get("mirror") | default("download.suse.de", true) }}/ibs/Devel:/Galaxy:/Manager:/4.0:/ToSLE/SLE_15_SP1/
```
Comment (moio) [inline salt/repos/repos.d/Devel_Galaxy_Manager_4.0_ToSLE.repo]: I guess this should also be added in `minima.yaml`, right?

File: salt/repos/repos.d/Devel_Galaxy_Manager_Head_ToSLE.repo [hunk 10]
```diff
@@ -0,0 +1,6 @@
+[Devel_Galaxy_Manager_Head_ToSLE]
+name=Devel Project for SUSE Manager Head (packages submitted directly to SLE)
+type=rpm-md
+enabled=1
+baseurl=http://{{ grains.get("mirror") | default("download.suse.de", true) }}/ibs/Devel:/Galaxy:/Manager:/Head:/ToSLE/SLE_15_SP1/
```
Comment (moio) [inline salt/repos/repos.d/Devel_Galaxy_Manager_Head_ToSLE.repo]: I guess this should also be added in `minima.yaml`, right?

File: salt/mirror/minima.yaml [hunk 11]
```diff
@@ -56,13 +56,26 @@ scc:
...
+    - SLE-Module-SUSE-Manager-Server-Pool
+    - SLE-Module-SUSE-Manager-Server-Updates
     # SUSE Manager Proxy
     - SUSE-Manager-Proxy-3.0-Pool
     - SUSE-Manager-Proxy-3.0-Updates
     - SUSE-Manager-Proxy-3.1-Pool
     - SUSE-Manager-Proxy-3.1-Updates
     - SUSE-Manager-Proxy-3.2-Pool
     - SUSE-Manager-Proxy-3.2-Updates
+    - SLE-Product-SUSE-Manager-Proxy-Pool
```
Comment (moio) [inline salt/mirror/minima.yaml]: I must have been sleeping under a rock but - what is the difference between those -Product and -Module repos?

File: salt/repos/repos.d/SUSE-Manager-4.0-x86_64-Pool.repo [hunk 12]
```diff
@@ -0,0 +1,6 @@
+[SUSE-Manager-4.0-x86_64-Pool]
+name=SUSE-Manager-4.0-x86_64-Pool
+type=rpm-md
+enabled=1
+baseurl=http://{{ grains.get("mirror") | default("download.suse.de", true) }}/ibs/Devel:/Galaxy:/Manager:/4.0/images/repo/SLE-Module-SUSE-Manager-Server-4.0-POOL-x86_64-Media1/
```
Comment (moio) [inline salt/repos/repos.d/SUSE-Manager-4.0-x86_64-Pool.repo]: I am not sure whether we want this repo. In other versions, what we did was to layer the non-image repo (`Devel_Galaxy_Manager_*.repo`) on top of the stable repo.

What is the reason to have both the image repo here and `Devel_Galaxy_Manager_*.repo` as well? When are we using which, and why, and why is it different from the past?

Also take care all repos should have an explicit priority value. Rationale is:

https://github.com/uyuni-project/sumaform/commit/d285f558b83b6baabd301365424195b088837143

and

https://github.com/uyuni-project/sumaform/commit/7392704d0b4ae32d6507c6d2b29bb8641a4a19d1

File: salt/repos/suse_manager_proxy.sls [hunk 13]
```diff
@@ -62,6 +88,14 @@ suse_manager_proxy_pool_repo:
     {% endif %}
     - template: jinja
 
+{% if grains['osfullname'] != 'Leap' %}
+suse_manager_devel_releasenotes_repo:
+  file.managed:
+    - name: /etc/zypp/repos.d/Devel_Galaxy_Manager_Head_ToSLE.repo
```
Comment (moio) [inline salt/repos/suse_manager_proxy.sls]: If we are sure this will always and however exclusively contain the release notes, I'd argue we should pick a file name that reflects that. Otherwise let's also make the state ID generic :wink:

**Comment by srbarrios** [comment]
Apart from moio questions, nothing else to comment :)

File: salt/repos/suse_manager_proxy.sls (hunk 13 above)
Comment (juliogonzalez) [reply salt/repos/suse_manager_proxy.sls]: Most probably this will include some more packages, such as the skelcd.

File: salt/repos/repos.d/SUSE-Manager-4.0-x86_64-Pool.repo (hunk 12 above)
Comment (juliogonzalez) [reply salt/repos/repos.d/SUSE-Manager-4.0-x86_64-Pool.repo]: The repository has a priority (line 97).

But the URL is a typo. This should be the module as this kind of `.repo` files are used for released version.

Fixing it.

File: salt/mirror/minima.yaml (hunk 5 above)
Comment (juliogonzalez) [reply salt/mirror/minima.yaml]: Repositories fixed now. Thanks for the hint!

File: modules/libvirt/pts/main.tf (hunk 7 above)
Comment (juliogonzalez) [reply modules/libvirt/pts/main.tf]: Rollback performed.

File: salt/repos/repos.d/Devel_Galaxy_Manager_4.0_ToSLE.repo (hunk 9 above)
Comment (juliogonzalez) [reply salt/repos/repos.d/Devel_Galaxy_Manager_4.0_ToSLE.repo]: Added.

File: salt/repos/repos.d/Devel_Galaxy_Manager_Head_ToSLE.repo (hunk 10 above)
Comment (juliogonzalez) [reply salt/repos/repos.d/Devel_Galaxy_Manager_Head_ToSLE.repo]: Added

File: salt/repos/repos.d/SUSE-Manager-4.0-x86_64-Pool.repo (hunk 12 above)
Comment (juliogonzalez) [reply salt/repos/repos.d/SUSE-Manager-4.0-x86_64-Pool.repo]: Fixed

File: salt/mirror/minima.yaml (hunk 8 above)
Comment (juliogonzalez) [reply salt/mirror/minima.yaml]: Fixed

File: README_ADVANCED.md (hunk 6 above)
Comment (juliogonzalez) [reply README_ADVANCED.md]: Fixed

File: salt/mirror/minima.yaml (hunk 11 above)
Comment (juliogonzalez) [reply salt/mirror/minima.yaml]: Product only contains the release notes and skelcd, and it comes from SLE directly (we submit there).

Modules contain the all the other packages, and comes from the SUSE Manager projects.

Reason: IIRC it's because why are now a base product.

**Comment by juliogonzalez** [comment]
After a small party with minima (see https://github.com/moio/minima/issues/14), I was able to verify that working with the mirror works as well for server and proxy.

Could not verify client tools for Devel, as we don't have them for 4.0 yet (3.2.8 MU needs to go out first), but we can merge this already and fix that later if needed (unlikely).
